sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala

```diff
@@ -39,40 +39,38 @@ object DeserializerBuildHelper {
       path: Expression,
       ordinal: Int,
       dataType: DataType,
-      walkedTypePath: Seq[String]): Expression = {
+      walkedTypePath: WalkedTypePath): Expression = {
     val newPath = GetStructField(path, ordinal)
     upCastToExpectedType(newPath, dataType, walkedTypePath)
   }
 
   def deserializerForWithNullSafety(
```

Comment:
so this is same as `expressionWithNullSafety`?

Reply:
Yeah now it's same. I've left comment on #23916. Btw I feel we might be better to focus review channel to one, either this PR or #23916 .

Reply:
I didn't notice you also make a lot change to `DeserializerBuildHelper` in this PR.  There might be conflicts if continuing #23916, I will close it.

Reply:
I think `expressionWithNullSafety` is more general naming so might be preferred one, but `deserializerForWithNullSafety` is also a good name cause we have relevant method `deserializerForWithNullSafetyAndUpcast`.

So that's a matter of preference and either can be removed. Which method would we prefer to keep?

Reply:
Thanks @viirya , please feel free to comment even it belongs to previous PR. Thanks again!

Reply:
Why you remove the `funcForCreatingNewExpr` from this and turn to pass in created expression (deserializer)?

Reply:
I tend to follow the suggestion (https://github.com/apache/spark/pull/23908#discussion_r261019295) unless I have strong opinion, as I'm fairly new to contributing SQL area. For consistency I agree having func is better, but for simplicity we can remove it like applying inline. Either is reasonable.

Reply:
I think we just need to keep `expressionWithNullSafety`. I don't see why we have to have 2 methods for `deserializeFor`. Leaving only a `deserializerForWithNullSafetyAndUpcast` is fine.

Reply:
Left `deserializerForWithNullSafetyAndUpcast` and `expressionWithNullSafety` since both are used in multiple places. Please let me know if it doesn't work.